### PR TITLE
Implement preliminary benchmark

### PR DESCRIPTION
**Pull request by kriskowal**
This is a preliminary benchmark that compares JSON and Thrift operations for arrays of coordinates, with various orders of magnitude. In short, ThriftRW can read substantially more quickly than JSON, and JSON is faster for writing.

The next benchmark will measure small structs.

```
p50  287ns p75  377ns p90  553ns stdev  128ns 0 json read
p50 17.4µs p75 22.5µs p90 29.4µs stdev 14.8µs 0 thrift read
p50  504ns p75  627ns p90  783ns stdev  237ns 0 json write pojo
p50  469ns p75  540ns p90  821ns stdev  288ns 0 thrift write pojo
p50  147ns p75  175ns p90  264ns stdev  213ns 0 thrift measure pojo
p50  604ns p75  718ns p90 1.11µs stdev  629ns 0 json write model
p50  461ns p75  533ns p90  723ns stdev  196ns 0 thrift write model
p50  146ns p75  169ns p90  235ns stdev 82.7ns 0 thrift measure model
p50  909ns p75 1.06µs p90 1.54µs stdev  434ns 1 json read
p50 16.1µs p75 20.9µs p90 26.4µs stdev 15.3µs 1 thrift read
p50 1.22µs p75 1.41µs p90 1.92µs stdev  370ns 1 json write pojo
p50 1.10µs p75 1.28µs p90 1.82µs stdev  438ns 1 thrift write pojo
p50  329ns p75  382ns p90  517ns stdev  117ns 1 thrift measure pojo
p50 1.24µs p75 1.46µs p90 2.02µs stdev  384ns 1 json write model
p50 1.21µs p75 1.29µs p90 1.94µs stdev  512ns 1 thrift write model
p50  325ns p75  376ns p90  471ns stdev  135ns 1 thrift measure model
p50 5.57µs p75 6.85µs p90 9.15µs stdev 1.59µs 10 json read
p50 16.5µs p75 21.9µs p90 28.4µs stdev 13.9µs 10 thrift read
p50 6.58µs p75 7.64µs p90 9.44µs stdev 1.74µs 10 json write pojo
p50 6.72µs p75 7.58µs p90 10.0µs stdev 1.79µs 10 thrift write pojo
p50 2.09µs p75 2.23µs p90 2.89µs stdev  551ns 10 thrift measure pojo
p50 6.68µs p75 7.70µs p90 9.99µs stdev 1.68µs 10 json write model
p50 6.54µs p75 7.07µs p90 8.40µs stdev 1.35µs 10 thrift write model
p50 2.04µs p75 2.15µs p90 2.55µs stdev  422ns 10 thrift measure model
p50 58.3µs p75 67.8µs p90 82.3µs stdev 12.6µs 100 json read
p50 18.6µs p75 24.3µs p90 29.2µs stdev 14.6µs 100 thrift read
p50 59.6µs p75 68.1µs p90 82.1µs stdev 11.6µs 100 json write pojo
p50 59.5µs p75 65.4µs p90 82.6µs stdev 10.9µs 100 thrift write pojo
p50 19.0µs p75 21.6µs p90 27.3µs stdev 4.54µs 100 thrift measure pojo
p50 60.8µs p75 69.5µs p90 87.8µs stdev 12.7µs 100 json write model
p50 73.5µs p75 77.2µs p90 89.0µs stdev 14.7µs 100 thrift write model
p50 18.8µs p75 21.3µs p90 25.1µs stdev 3.71µs 100 thrift measure model
p50  581µs p75  639µs p90  678µs stdev 60.9µs 1000 json read
p50 17.2µs p75 23.4µs p90 29.7µs stdev 14.5µs 1000 thrift read
p50  735µs p75  804µs p90  932µs stdev 92.7µs 1000 json write pojo
p50  640µs p75  688µs p90  740µs stdev 58.8µs 1000 thrift write pojo
p50  178µs p75  198µs p90  221µs stdev 24.2µs 1000 thrift measure pojo
p50  684µs p75  726µs p90  761µs stdev 45.1µs 1000 json write model
p50  636µs p75  663µs p90  763µs stdev 62.4µs 1000 thrift write model
p50  184µs p75  201µs p90  226µs stdev 25.0µs 1000 thrift measure model
p50 5.86ms p75 5.90ms p90 6.12ms stdev  132µs 10000 json read
p50 17.7µs p75 24.0µs p90 30.0µs stdev 14.3µs 10000 thrift read
p50 12.6ms p75 12.6ms p90 12.7ms stdev 60.8µs 10000 json write pojo
p50 6.10ms p75 6.22ms p90 6.23ms stdev 77.5µs 10000 thrift write pojo
p50 2.02ms p75 2.08ms p90 2.67ms stdev  276µs 10000 thrift measure pojo
p50 12.7ms p75 13.1ms p90 13.4ms stdev  386µs 10000 json write model
p50 6.18ms p75 6.19ms p90 6.29ms stdev  119µs 10000 thrift write model
p50 1.91ms p75 2.00ms p90 2.02ms stdev  113µs 10000 thrift measure model
```

### CODE REVIEW

**Comment by kriskowal** [comment]
r @Raynos @rf 
cc @jcorbin @jwolski @malandrew @Matt-Esch

File: benchmarks/array-of-coordinates.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        coordinates: unstructuredCoordinates
+    };
+
+    var lengthRes = thrift.CoordinatesHolder.rw.byteLength(structuredEnvelope);
+    var length = lengthRes.length;
+    var buffer = new Buffer(length);
+    var json = JSON.stringify(unstructuredEnvelope);
+
+    var cases = {
+        'json read': function b() {
```
Comment (Raynos) [inline benchmarks/array-of-coordinates.js]: Please name each one uniquely here so we can see which are deopted.

File: benchmarks/array-of-coordinates.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            JSON.parse(json);
+        },
+        'thrift read': function b() {
+            thrift.CoordinatesHolder.rw.readFrom(buffer, 0);
+        },
+        'json write pojo': function b() {
+            JSON.stringify(unstructuredEnvelope);
```
Comment (Raynos) [inline benchmarks/array-of-coordinates.js]: We are calling a closure in a loop. Are we sure this is a good approach ?

I feel like having no closures makes the benchmark more reliable.

**Comment by kriskowal** [comment]
I’ve updated this benchmark based on @raynos’s feedback and decided after a conversation with @matt-esch to compare two scenarios, JSON and Thrift, where both do a round trip through a buffer and do validation, in this case using JSEN for the JSON validation, equivalent to what Thrift does inherently.

```
p50 1.78µs p75 2.10µs p90 2.99µs stdev  644ns 0 thrift
p50 2.76µs p75 3.25µs p90 4.62µs stdev  926ns 0 json
p50 3.81µs p75 4.48µs p90 6.08µs stdev 1.73µs 1 thrift
p50 4.88µs p75 5.50µs p90 7.00µs stdev 1.74µs 1 json
p50 21.1µs p75 24.6µs p90 31.3µs stdev 6.25µs 10 thrift
p50 15.5µs p75 18.0µs p90 22.5µs stdev 4.06µs 10 json
p50  192µs p75  215µs p90  242µs stdev 29.9µs 100 thrift
p50  137µs p75  161µs p90  182µs stdev 23.5µs 100 json
p50 2.08ms p75 2.16ms p90 2.19ms stdev  169µs 1000 thrift
p50 1.48ms p75 1.59ms p90 1.72ms stdev  129µs 1000 json
p50 20.6ms p75 20.6ms p90 20.9ms stdev  339µs 10000 thrift
p50 19.6ms p75 19.6ms p90 20.1ms stdev  533µs 10000 json
```

**Comment by kriskowal** [comment]
I’ve added a benchmark for single objects with 24 optional string fields. The bench varies the number of provided fields and compares to the equivalent JSON/JSEN combo. This case does not look as good.

```
p25 48.5µs p50 67.0µs p75 99.0µs p90  152µs stddev 9.24µs q3-q1/2  104µs 0 thrift
p25 1.77µs p50 2.04µs p75 2.58µs p90 3.52µs stddev  134ns q3-q1/2 1.29µs 0 json
p25 59.2µs p50 64.5µs p75 91.1µs p90  118µs stddev 2.69µs q3-q1/2 30.3µs 6 thrift
p25 4.56µs p50 5.54µs p75 8.21µs p90 11.8µs stddev  488ns q3-q1/2 8.94µs 6 json
p25 45.7µs p50 53.7µs p75 75.3µs p90  108µs stddev 4.01µs q3-q1/2 30.5µs 12 thrift
p25 5.50µs p50 6.15µs p75 7.20µs p90 9.56µs stddev  328ns q3-q1/2 1.81µs 12 json
p25 46.6µs p50 54.1µs p75 74.5µs p90  105µs stddev 3.73µs q3-q1/2 30.4µs 18 thrift
p25 9.58µs p50 11.0µs p75 13.9µs p90 19.1µs stddev  723ns q3-q1/2 10.0µs 18 json
```

**Pull request merged**